### ISSUE TITLE
ENH 

### ISSUE THREAD

**Issue by JacekPliszka** [pandas-dev/pandas#20521]
#### Code Sample, a copy-pastable example if possible

Please comsider merging

https://github.com/JacekPliszka/pandas/commit/9459bdfabf3eff38da77718b9ec77e86b0ede54a

#### Problem description

Currently pandas can not add custom metadata to parquet file.

This patch add metadata argument to DataFrame.to_parquet that allows for that.
Warning is issued when pandas key is present in the dictionary passed.

**Comment by TomAugspurger** [comment]
cc @cpcloud 

What's the purpose here? Would this be in addition to or in place of the usual `pandas_metadata`?

**Comment by JacekPliszka** [comment]
The user given dictionary updates current key value file metadata. If user gives pandas key then it overwrites pandas_metadata but warning.warn is issued.

Purpose:

User metadata is very needed when:

1. processing is done in several stages and you want to keep information about version/algorithm used on each stage so you can debug it later

2. processing is done with different parameters and you want to keep parameters used with the file

3. you need to add extra custom information e.g. sometimes one column comes from one source and sometimes it is calculated from other columns  and you want to keep this information and pass to later stages of processing

4. you have certain very high level aggregates that are costly to compute and you do not want to create columns for them

For me it is very important features and one of the main reasons I want to switch to parquet.

**Comment by TomAugspurger** [comment]
That all sounds reasonable.

**Comment by JacekPliszka** [comment]
And added whatsnew and rebased to current master.

**Comment by jorisvandenbossche** [comment]
Note for readers: the PR was closed but mentions a work-around that can be used for now if you need this: https://github.com/pandas-dev/pandas/pull/20534#issuecomment-453236538

**Comment by snowman2** [comment]
I have been thinking about this and am wondering what the general thoughts are to use [DataFrame.attrs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.attrs.html#pandas-dataframe-attrs) and [Series.attrs](https://pandas.pydata.org/pandas-docs/stable//reference/api/pandas.Series.attrs.html#pandas-series-attrs) for reading and writing metadata to/from parquet?

For example, here is how the metadata would be written:
```python
pdf = pandas.DataFrame({"a": [1]})
pdf.attrs = {"name": "my custom dataset"}
pdf.a.attrs = {"long_name": "Description about data", "nodata": -1, "units": "metre"}
pdf.to_parquet("file.parquet")
```
Then, when loading in the data:
```python
pdf = pandas.read_parquet("file.parquet")
pdf.attrs
```
```
{"name": "my custom dataset"}
```
```python
pdf.a.attrs
```
```
{"long_name": "Description about data", "nodata": -1, "units": "metre"}
```
Is this something that would need to be done in pandas or pyarrow/fastparquet?

**Comment by snowman2** [comment]
Here is a hack to get the attrs to work with pyarrow:
```python
def to_parquet(pdf, filename):
    # write parquet file with attributes
    table = pyarrow.Table.from_pandas(pdf)
    schema_metadata = dict(table.schema.metadata or {})
    pandas_metadata = json.loads(schema_metadata.get(b"pandas", {}))   
    pandas_metadata.update(
        attrs=json.dumps(pdf.attrs),
        column_attrs=json.dumps({col: pdf[col].attrs for col in pdf.columns})
    )
    schema_metadata[b"pandas"] = json.dumps(pandas_metadata)
    table = table.replace_schema_metadata(schema_metadata)
    pyarrow.parquet.write_table(table, filename)

def read_parquet(filename):
    # read parquet file with attributes
    table = pyarrow.parquet.read_pandas(filename)
    pandas_metadata = json.loads(table.schema.metadata[b"pandas"])
    pdf = table.to_pandas()
    pdf.attrs = json.loads(pandas_metadata.get("attrs", '{}'))
    col_attrs = json.loads(pandas_metadata.get("column_attrs", '{}'))
    for col, attrs in col_attrs.items():
        pdf[col].attrs = attrs
    return pdf
```

Example:

Writing:
```python
pdf = pandas.DataFrame({"a": [1]})
pdf.attrs = {"name": "my custom dataset"}
pdf.a.attrs = {"long_name": "Description about data", "nodata": -1, "units": "metre"}
to_parquet(pdf, "a.parquet")
```
Reading:
```python
pdf = read_parquet("a.parquet")
pdf.attrs
```
```
{"name": "my custom dataset"}
```
```python
pdf.a.attrs
```
```
{"long_name": "Description about data", "nodata": -1, "units": "metre"}
```

**Comment by snowman2** [comment]
I have a PR that seems to do the trick: #41545

**Comment by jorisvandenbossche** [comment]
> Is this something that would need to be done in pandas or pyarrow/fastparquet?

Ideally, I think this would actually be done in pyarrow/fastparquet, as it is in those libraries that the "pandas" metadata item gets constructed currently

**Comment by gsheni** [comment]
What if we added an argument `custom_metadata` to the `to_parquet` function? end-users could add any key value pair they wanted, so long as there wasn't a `pandas` key.

**Comment by arogozhnikov** [comment]
so... can we have simple something to work with df.attrs ?

The goal is to replace multiple pseudo-csv formats which add #-prefixed comments in the beginning of a file with something systematic.

I believe everyone would agree that's 1) a common usecase 2) supportable by parquet 3) should work without hassle for reader (I'm ok with hassle for writer)

**Comment by jorisvandenbossche** [comment]
> I believe everyone would agree that's 1) a common usecase 2) supportable by parquet 3) should work without hassle for reader (I'm ok with hassle for writer)

Yes, and a contribution to add this functionality is welcome, I think. 
https://github.com/pandas-dev/pandas/pull/41545 tried to do this but was only closed because it also wanted to store column-level attrs (which was the main driver for the PR author), not because we don't want this in general. A PR focusing on storing/restoring DataFrame-level attrs is welcome.

And a PR to add generic parquet file-level metadata with a `metadata` keyword (as was attempted in  #20534, and the original purpose of this issue) is also still welcome I think.

**Comment by davetapley** [comment]
My workaround (assuming `fastparquet`):

```py
# write
df.to_parquet(path)
meta = {'foo':'bar'}
fastparquet.update_file_custom_metadata(path, meta)

# read
pf = fastparquet.ParquetFile(path)
df_ = pf.to_pandas()
meta_ = pf.key_value_metadata
```

Note `meta` **must** be `dict[str, str]` (so no nested dicts without bring-your-own serialization).

**Comment by davetapley** [comment]
This is done and in `2.1.0` 🎉 
- https://github.com/pandas-dev/pandas/pull/54346
